relayer/src/chain/cosmos/estimate.rs

```diff
@@ -0,0 +1,147 @@
+use ibc::core::ics24_host::identifier::ChainId;
+use ibc_proto::cosmos::tx::v1beta1::{Fee, Tx};
+use ibc_proto::google::protobuf::Any;
+use tonic::codegen::http::Uri;
+use tracing::{debug, error, span, warn, Level};
+
+use crate::chain::cosmos::encode::sign_tx;
+use crate::chain::cosmos::gas::{gas_amount_to_fees, PrettyFee};
+use crate::chain::cosmos::simulate::send_tx_simulate;
+use crate::chain::cosmos::types::account::Account;
+use crate::chain::cosmos::types::gas::GasConfig;
+use crate::config::types::Memo;
+use crate::config::ChainConfig;
+use crate::error::Error;
+use crate::keyring::KeyEntry;
+
+pub async fn estimate_tx_fees(
+    config: &ChainConfig,
+    grpc_address: &Uri,
+    key_entry: &KeyEntry,
+    account: &Account,
+    tx_memo: &Memo,
+    messages: Vec<Any>,
+) -> Result<Fee, Error> {
+    let gas_config = GasConfig::from_chain_config(config);
+
+    debug!(
+        "max fee, for use in tx simulation: {}",
+        PrettyFee(&gas_config.max_fee)
+    );
+
+    let signed_tx = sign_tx(
+        config,
+        key_entry,
+        account,
+        tx_memo,
+        messages,
+        &gas_config.max_fee,
+    )?;
+
+    let tx = Tx {
+        body: Some(signed_tx.body),
+        auth_info: Some(signed_tx.auth_info),
+        signatures: signed_tx.signatures,
+    };
+
+    let estimated_fee = estimate_fee_with_tx(&gas_config, grpc_address, &config.id, tx).await?;
+
+    Ok(estimated_fee)
+}
+
+async fn estimate_fee_with_tx(
```

Comment:
This is refactored from:

https://github.com/informalsystems/ibc-rs/blob/3e34b4b916eb91f680412e5fdd3ea045a99526d2/relayer/src/chain/cosmos.rs#L332-L348